feat: add pagination to _getRailsForAddressAndToken

DarkLord017 opened the pull request:
Resolves #216 
 introduces pagination support to the functions that fetch rails  The main changes involve updating the public API to accept `offset` and `limit` parameters, modifying the internal logic to handle pagination, and returning additional metadata to support client-side pagination.

wjmelements commented on src/Payments.sol:
```diff
@@ -1686,19 +1686,24 @@ contract Payments is ReentrancyGuard {
-        returns (RailInfo[] memory)
+        returns (RailInfo[] memory , uint256 nextOffset , uint256 total)
     {
+        if (offset >= rails.length) return (new RailInfo[](0), rails.length, rails.length);
```
weird to check `rails.length` instead of `allRailIds.length`

DarkLord017 commented:
* Updated the external functions as asked . 
* Added a new `testPagination` function in `RailGetters.t.sol` to test the pagination logic for payer and payee , and fixed other tests to accomodate new syntax

please review @wjmelements

wjmelements commented on src/Payments.sol:
```diff
@@ -1663,42 +1663,57 @@ contract Payments is ReentrancyGuard {
-    function getRailsForPayerAndToken(address payer, IERC20 token) external view returns (RailInfo[] memory) {
-        return _getRailsForAddressAndToken(payer, token, true);
+    function getRailsForPayerAndToken(address payer, IERC20 token , uint256 offset , uint256 limit) external view returns (RailInfo[] memory results, uint256 nextOffset , uint256 total) {
+        if (offset >= payerRails[token][payer].length) return (new RailInfo[](0), payerRails[token][payer].length, payerRails[token][payer].length);
```
don't do this here; do it in `_getRailsForAddressAndToken`

wjmelements commented on src/Payments.sol:
```diff
@@ -1663,42 +1663,57 @@ contract Payments is ReentrancyGuard {
         uint256 railsLength = allRailIds.length;
 
-        RailInfo[] memory results = new RailInfo[](railsLength);
+        results = new RailInfo[](railsLength);
```
shouldn't the length of results be `limit`?

wjmelements replied:
or rather, `end - offset`

wjmelements commented on src/Payments.sol:
```diff
@@ -1663,42 +1663,57 @@ contract Payments is ReentrancyGuard {
 
-        RailInfo[] memory results = new RailInfo[](railsLength);
+        results = new RailInfo[](railsLength);
         uint256 resultCount = 0;
```
nope I am wrong:
```
            // Skip non-existent rails
            if (rail.from == address(0)) continue;
```

DarkLord017 commented:
please review @wjmelements

wjmelements commented on src/Payments.sol:
```diff
@@ -1663,42 +1663,63 @@ contract Payments is ReentrancyGuard {
     {
-        // Get the appropriate list of rails based on whether we're looking for payer or payee
-        uint256[] storage allRailIds = isPayer ? payerRails[token][addr] : payeeRails[token][addr];
+        if (offset >= allRailIds.length) return (new RailInfo[](0), allRailIds.length, allRailIds.length);
```
do this after getting railsLength.

wjmelements commented:
When writing solidity, you want to first minimize storage accesses like `.length`.

Many developers are used to developing in languages with decent compiler optimization. Pretend the solc compiler does not have optimization; in many cases the compiler optimization is worse than you imagine.

DarkLord017 commented:
@wjmelements fixed the issue
